### PR TITLE
Peter/base rest api

### PR DESCRIPTION
**Pull request by ndersorn**
Dette er et draft af 
Webapi Implementation of User(+Creation), Home(+Creation) and Hub(+Creation,GetByHome && GetByUser)

Ved ikke om der er mere der skal laves hertil?

### CODE REVIEW

File: source/cloudserver/webapi/XPowerApi/XPowerApi/Managers/TokenManager.cs [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+                    ValidateAudience = false,
+                    IssuerSigningKey = new SymmetricSecurityKey(key)
+                };
+                SecurityToken securityToken;
+                ClaimsPrincipal principal = tokenHandler.ValidateToken(token,
+                      parameters, out securityToken);
+                return principal;
+            }
+            catch
+            {
```
Comment (Ripiter) [inline source/cloudserver/webapi/XPowerApi/XPowerApi/Managers/TokenManager.cs:110]: can we create a log here?

File: source/cloudserver/webapi/XPowerApi/XPowerApi/Models/HubModels/HubCreate.cs [hunk 2]
```diff
@@ -0,0 +1,14 @@
+namespace XPowerApi.Models.HubModels;
+
+public class HubCreate
+{
+    public string Name { get; set; }
+
+    public string Mac { get; set; }
+    
+    public string Home { get; set; }
```
Comment (Ripiter) [inline source/cloudserver/webapi/XPowerApi/XPowerApi/Models/HubModels/HubCreate.cs:9]: when creating a hub, do we know the name of it and in what home it belongs?

File: source/cloudserver/webapi/XPowerApi/XPowerApi/Models/UserModels/UserCreate.cs [hunk 3]
```diff
@@ -0,0 +1,8 @@
+﻿namespace XPowerApi.Models.UserModels
+{
+    public class UserCreate
+    {
+        public string UserName { get; set; }
```
Comment (Ripiter) [inline source/cloudserver/webapi/XPowerApi/XPowerApi/Models/UserModels/UserCreate.cs:5]: Add email, first name and last name

File: source/cloudserver/webapi/XPowerApi/XPowerApi/Providers/HomeProvider.cs [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+            _dbProvider = new SurrealDbProvider(configuration);
+        }
+
+        public async Task<Home> CreateHome(HomeCreate homeCreate)
+        {
+            // Create HomeGroup DB Object
+            Dictionary<string, string> dataArray = new Dictionary<string, string>()
+                { { "name", homeCreate.Name }, };
+            // Create Db object and Convert to HomeGroup on success
+            Home home = (await _dbProvider.Create<HomeDb>("home", dataArray)).ConvertToHome();
```
Comment (Ripiter) [inline source/cloudserver/webapi/XPowerApi/XPowerApi/Providers/HomeProvider.cs:25]: converting in manager

File: source/cloudserver/webapi/XPowerApi/XPowerApi/Providers/HomeProvider.cs [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+            return home;
+        }
+
+        public async Task<Home> GetHomeById(int id)
+        {
+            Home home = (await _dbProvider.GetOneById<HomeDb>("home", id)).ConvertToHome();
```
Comment (Ripiter) [inline source/cloudserver/webapi/XPowerApi/XPowerApi/Providers/HomeProvider.cs:31]: converting in manager

File: source/cloudserver/webapi/XPowerApi/XPowerApi/Providers/HubProvider.cs [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+            _dbProvider = new SurrealDbProvider(configuration);
+        }
+
+        public async Task<Hub> CreateHub(HubCreate hubCreate)
+        {
+            // Create HomeGroup DB Object
+            Dictionary<string, string> dataArray = new Dictionary<string, string>()
+                { { "name", hubCreate.Name }, };
+            // Create Db object and Convert to HomeGroup on success
+            Hub hub = (await _dbProvider.Create<HubDb>("hub", dataArray)).ConvertToHub();
```
Comment (Ripiter) [inline source/cloudserver/webapi/XPowerApi/XPowerApi/Providers/HubProvider.cs:26]: converting in manager

File: source/cloudserver/webapi/XPowerApi/XPowerApi/Providers/HubProvider.cs [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+            return hub;
+        }
+
+        public async Task<Hub> GetHubById(int id)
+        {
+            Hub hub = (await _dbProvider.GetOneById<HubDb>("home", id)).ConvertToHub();
```
Comment (Ripiter) [inline source/cloudserver/webapi/XPowerApi/XPowerApi/Providers/HubProvider.cs:32]: converting in manager

File: source/cloudserver/webapi/XPowerApi/XPowerApi/Providers/HubProvider.cs [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+            return hub;
+        }
+        
+        public async Task<List<Hub>> GetHubsByHomeId(int homeid)
+        {
+            List<Hub> hubs = new List<Hub>();
+            List<HubDb> dbHubList = (await _dbProvider.GetOneFromInsideAnother<HubDb>("hub", "home", $"home:{homeid}"));
+            foreach (var hubDb in dbHubList)
```
Comment (Ripiter) [inline source/cloudserver/webapi/XPowerApi/XPowerApi/Providers/HubProvider.cs:40]: converting in manager

File: source/cloudserver/webapi/XPowerApi/XPowerApi/Providers/UserProvider.cs [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+
+            // Create User DB Object
+            Dictionary<string, string> dataArray = new Dictionary<string, string>()
+            {
+                { "hashed_password", hashed },
+                { "username", userCreate.UserName },
+                { "salt", System.Text.Encoding.UTF8.GetString(salt) }
+            };
+            // Create Db object and Convert to User on success
+            User user = (await _dbProvider.Create<UserDb>("user", dataArray)).ConvertToUser();
```
Comment (Ripiter) [inline source/cloudserver/webapi/XPowerApi/XPowerApi/Providers/UserProvider.cs:38]: converting in manager

File: source/cloudserver/webapi/XPowerApi/XPowerApi/Providers/UserProvider.cs [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+
+        public async Task<User> GetUserById(int id)
+        {
+            User user = (await _dbProvider.GetOneById<UserDb>("user", id)).ConvertToUser();
+            return user;
+        }
+
+        public Task<User> UpdateUserUsername(int id, string username)
+        {
+            throw new NotImplementedException();
```
Comment (Ripiter) [inline source/cloudserver/webapi/XPowerApi/XPowerApi/Providers/UserProvider.cs:50]: not implemented, remove from interface or implement it

**Review by Ripiter — CHANGES_REQUESTED** [review]
Added some comments

File: source/cloudserver/webapi/XPowerApi/XPowerApi/Controllers/HubController.cs [hunk 11]
```diff
@@ -0,0 +1,116 @@
+using Microsoft.AspNetCore.Mvc;
+using XPowerApi.Interfaces;
+using XPowerApi.Models.HubModels;
+
+namespace XPowerApi.Controllers;
```
[the remark this answers is not in the document]
Comment (ndersorn) [reply source/cloudserver/webapi/XPowerApi/XPowerApi/Controllers/HubController.cs:5]: Forstår ikke hvad du mener?

File: source/cloudserver/webapi/XPowerApi/XPowerApi/Controllers/UserCreationController.cs [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+                return BadRequest(new { Message = "Username or Password is missing valid" });
+
+            try
+            {
+                //Checking the user
+                var user = await _userManager.CreateUser(userCreateInfo);
+                //Checking the homegroup
+                var home = await _homeManager.CreateHome(new HomeCreate() { Name = "Default" });
+                //make relation
+                var related = await _homeManager.RelateUserToHome(user.Id, home.Id);
```
[the remark this answers is not in the document]
Comment (ndersorn) [reply source/cloudserver/webapi/XPowerApi/XPowerApi/Controllers/UserCreationController.cs:57]: Incomming

File: source/cloudserver/webapi/XPowerApi/XPowerApi/DbModels/SurrealDbModels/RelateObject.cs [hunk 13]
```diff
@@ -0,0 +1,17 @@
...
+
+public class RelateObject
+{
+    public string Id { get; set; }
+    
+    public string In { get; set; }
+    
+    public string Out { get; set; }
+
+    public RelateObject(string Id, string In, string Out)
```
[the remark this answers is not in the document]
Comment (ndersorn) [reply source/cloudserver/webapi/XPowerApi/XPowerApi/DbModels/SurrealDbModels/RelateObject.cs:11]: was on purpose, can't go lowerCase, "in" and "out" is reserved keywords.

File: source/cloudserver/webapi/XPowerApi/XPowerApi/Interfaces/IDbProvider.cs [hunk 14]
```diff
@@ -0,0 +1,16 @@
+namespace XPowerApi.Interfaces;
+
+public interface IDbManager
+{
+    public Task<T> Create<T>(string tableName, Dictionary<string, string> dataArray) where T : new();
+
+    public Task<int> GetNextId(string tableName);
+    public Task<T> GetOneById<T>(string tableName, int id);
+    public Task<T> GetSpecific<T>(string tableName, string whereClause = "");
```
[the remark this answers is not in the document]
Comment (ndersorn) [reply source/cloudserver/webapi/XPowerApi/XPowerApi/Interfaces/IDbProvider.cs:9]: incomming

File: source/cloudserver/webapi/XPowerApi/XPowerApi/Managers/TokenManager.cs (hunk 1 above)
Comment (ndersorn) [reply source/cloudserver/webapi/XPowerApi/XPowerApi/Managers/TokenManager.cs:110]: Made by @blackout1471, I wouldn't know how to atm.

File: source/cloudserver/webapi/XPowerApi/XPowerApi/Models/HubModels/HubCreate.cs (hunk 2 above)
Comment (ndersorn) [reply source/cloudserver/webapi/XPowerApi/XPowerApi/Models/HubModels/HubCreate.cs:9]: i would assume yes?

File: source/cloudserver/webapi/XPowerApi/XPowerApi/Models/UserModels/UserCreate.cs (hunk 3 above)
Comment (ndersorn) [reply source/cloudserver/webapi/XPowerApi/XPowerApi/Models/UserModels/UserCreate.cs:5]: incomming

File: source/cloudserver/webapi/XPowerApi/XPowerApi/Providers/HomeProvider.cs (hunk 4 above)
Comment (ndersorn) [reply source/cloudserver/webapi/XPowerApi/XPowerApi/Providers/HomeProvider.cs:25]: ?

File: source/cloudserver/webapi/XPowerApi/XPowerApi/Providers/SurrealDbProvider.cs [hunk 15]
```diff
@@ -0,0 +1,141 @@
...
+    {
+        string sqlString = $"select * from ${tableName} where ${baseTable} inside (select out as id from ${relationTable} where in is ${targetId});";
+        SurrealDbResult dbResult = await MakeRawResult(sqlString);
+        List<T> objectList = JsonSerializer.Deserialize<List<T>>(JsonSerializer.Serialize(dbResult.result[0]))!;
+        return objectList;
+    }
+
+    public async Task<T> Update<T>(string tableName, int id, T newT)
+    {
+        throw new NotImplementedException();
```
[the remark this answers is not in the document]
Comment (ndersorn) [reply source/cloudserver/webapi/XPowerApi/XPowerApi/Providers/SurrealDbProvider.cs:134]: removed

File: source/cloudserver/webapi/XPowerApi/XPowerApi/Providers/UserProvider.cs (hunk 10 above)
Comment (ndersorn) [reply source/cloudserver/webapi/XPowerApi/XPowerApi/Providers/UserProvider.cs]: fixed

File: source/cloudserver/webapi/XPowerApi/XPowerApi/Controllers/UserCreationController.cs [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+                string.IsNullOrEmpty(userCreateInfo.LastName)
+            )
+                return BadRequest(new { Message = "Username or Password is missing valid" });
+
+            try
+            {
+                //Checking the user
+                var user = await _userManager.CreateUser(userCreateInfo);
+                //Checking the homegroup
+                var home = await _homeManager.CreateHome(new HomeCreate() { Name = "Default" }, user.Id);
```
Comment (Ripiter) [inline source/cloudserver/webapi/XPowerApi/XPowerApi/Controllers/UserCreationController.cs:58]: why not take name as parameter and new HomeCreate in the method?

Comment (ndersorn) [reply source/cloudserver/webapi/XPowerApi/XPowerApi/Controllers/UserCreationController.cs:58]: Under UserCreate, snakkede vi om altid at være i en Default gruppe, så det er min begrundelse for dette.

File: source/cloudserver/webapi/XPowerApi/XPowerApi/Managers/UserManager.cs [hunk 17]
```diff
@@ -0,0 +1,36 @@
...
+        public async Task<User> CreateUser(UserCreate userCreate)
+        {
+            // Generate Salt
+            byte[] salt = SecuritySupport.GenerateSalt();
+            string hashed_password = SecuritySupport.HashPassword(userCreate.Password!, salt);
+            // Create User DB Object
+            Dictionary<string, string> dataArray = new Dictionary<string, string>()
+            {
+                { "hashed_password", hashed_password },
+                { "username", userCreate.UserName },
```
Comment (Ripiter) [inline source/cloudserver/webapi/XPowerApi/XPowerApi/Managers/UserManager.cs:25]: first name, lastname not set in db

Comment (ndersorn) [reply source/cloudserver/webapi/XPowerApi/XPowerApi/Managers/UserManager.cs:25]: my bad

File: source/cloudserver/webapi/XPowerApi/XPowerApi/Controllers/UserCreationController.cs (hunk 16 above)
Comment (Ripiter) [reply source/cloudserver/webapi/XPowerApi/XPowerApi/Controllers/UserCreationController.cs:58]: i explained it poorly, here is what i meant

class HomeManager
{
    Home CreateHome(string name, int userID)
   {
      var newHome = new HomeCreate() { Name = name };
    }
}

Comment (ndersorn) [reply source/cloudserver/webapi/XPowerApi/XPowerApi/Controllers/UserCreationController.cs:58]: Det er også basiclly det der bliver gjort i HomeManager.CreateHome()
Bare med HomeCreate Objectet, for at kunne skallerer objected hvis nødvendigt.
Og så er det consistent med de andre CreationControllers

File: source/cloudserver/webapi/XPowerApi/XPowerApi/Controllers/HomeCreationController.cs [hunk 18]
```diff
@@ -0,0 +1,56 @@
...
+        [HttpPost]
+        [Route("CreateHome")]
+        public async Task<ActionResult<Home>> CreateHome(HomeCreate homeCreateInfo, int userId)
+        {
+            if (homeCreateInfo == null)
+                return BadRequest(new { Message = "Missing parameter" });
+
+            //checking if name is not null or empty
+            if (string.IsNullOrEmpty(homeCreateInfo.Name))
+                return BadRequest(new { Message = "Name is missing valid" });
```
Comment (Ripiter) [inline source/cloudserver/webapi/XPowerApi/XPowerApi/Controllers/HomeCreationController.cs:34]: Ah yess the good old english

File: source/cloudserver/webapi/XPowerApi/XPowerApi/Controllers/HubCreationController.cs [hunk 19]
```diff
@@ -0,0 +1,57 @@
...
+        [Route("CreateHub")]
+        public async Task<ActionResult<Hub>> CreateHub(HubCreate hubCreateInfo)
+        {
+            if (hubCreateInfo == null)
+                return BadRequest(new { Message = "Missing parameter" });
+
+            //checking if Home, Mac_addr and Pricate_addr is not null or empty
+            if (string.IsNullOrEmpty(hubCreateInfo.Home) || string.IsNullOrEmpty(hubCreateInfo.Mac) ||
+                string.IsNullOrEmpty(hubCreateInfo.PrivateAddress))
+                return BadRequest(new { Message = "Name, Mac_addr or PrivateAddress is missing valid" });
```
Comment (Ripiter) [inline source/cloudserver/webapi/XPowerApi/XPowerApi/Controllers/HubCreationController.cs:35]: remove valid at the end

File: source/cloudserver/webapi/XPowerApi/XPowerApi/Controllers/UserCreationController.cs [hunk 20]
```diff
@@ -0,0 +1,75 @@
...
+
+            //checking if username and password is not null or empty
+            if (
+                string.IsNullOrEmpty(userCreateInfo.UserName) ||
+                string.IsNullOrEmpty(userCreateInfo.Password) ||
+                string.IsNullOrEmpty(userCreateInfo.Email) ||
+                string.IsNullOrEmpty(userCreateInfo.FirstName) ||
+                string.IsNullOrEmpty(userCreateInfo.LastName)
+            )
+                return BadRequest(new { Message = "Username or Password is missing valid" });
```
Comment (Ripiter) [inline source/cloudserver/webapi/XPowerApi/XPowerApi/Controllers/UserCreationController.cs:51]: remove valid

**Review by Ripiter — CHANGES_REQUESTED** [review]
grammar

**Review by Ripiter — APPROVED** [review]
looks good

**Pull request merged**